In [1]:
# import libraries
import pandas as pd
import numpy as np 
import matplotlib.pyplot as plt
import datetime
from sklearn import preprocessing, linear_model, model_selection
from sklearn.tree import DecisionTreeRegressor as DTR
import statsmodels.api as sm
from itertools import product

%matplotlib inline

/home/frodos/anaconda/lib/python2.7/site-packages/statsmodels/compat/pandas.py:56: FutureWarning: The pandas.core.datetools module is deprecated and will be removed in a future version. Please use the pandas.tseries module instead.
  from pandas.core import datetools


Класс моделей ARIMA недостаточно богат для наших данных: с их помощью, например, никак нельзя учесть взаимосвязи между рядами. Это можно сделать с помощью векторной авторегрессии VARIMA, но её питоновская реализация не позволяет использовать регрессионные признаки. Кроме того, авторегрессионный подход не позволяет учитывать, например, взаимодействия между сезонными компонентами. Вы могли заметить, что форма суточных сезонных профилей в будни и выходные немного разная; явно моделировать этот эффект с помощью ARIMA не получится.

Нам нужна более сложная модель. Давайте займёмся сведением задачи массового прогнозирования рядов к регрессионной постановке! Вам понадобится много признаков. Некоторые из них у вас уже есть — это:
<ol>
<li>идентификатор географической зоны
<li>дата и время
<li>количество поездок в периоды, предшествующие прогнозируемому
<li>синусы, косинусы и тренды, которые вы использовали внутри регрессионной компоненты ARIMA
<li>Кроме того, не спешите выбрасывать построенный вами на прошлой неделе прогнозы — из них может получиться хороший признак для регрессии!
</ol>



Вы можете попробовать разные регрессионный модели, но хорошие результаты, скорее всего, дадут такие, которые будут позволять признакам взаимодействовать друг с другом.

Поскольку прогноз нужен на 6 часов вперёд, проще всего будет построить 6 независимых регрессионных моделей — одна для прогнозирования y^T+1|T, другая для y^T+2|T и т.д.

<ol>Чтобы сдать задание, выполните следующую последовательность действий.
<li>Для каждой из шести задач прогнозирования y^T+i|T,i=1,…,6 сформируйте выборки. Откликом будет yT+i при всевозможных значениях T, а признаки можно использовать следующие:
<ul>
<li>идентификатор географической зоны — категориальный
<li>год, месяц, день месяца, день недели, час — эти признаки можно пробовать брать и категориальными, и непрерывными, можно даже и так, и так (done)
<li>синусы, косинусы и тренды, которые вы использовали внутри регрессионной компоненты ARIMA (done)
<li>сами значения прогнозов ARIMA y^T+i|TARIMA
<li>количество поездок из рассматриваемого района в моменты времени yT,yT−1,…,yT−K (параметр K можно подбирать; попробуйте начать, например, с 6)
<li>количество поездок из рассматриваемого района в моменты времени yT−24,yT−48,…,yT−24∗Kd (параметр Kd можно подбирать; попробуйте начать, например, с 2)
<li>суммарное количество поездок из рассматриваемого района за предшествующие полдня, сутки, неделю, месяц
</ul>
Будьте внимательны при создании признаков — все факторы должны быть рассчитаны без использования информации из будущего: при прогнозировании y^T+i|T,i=1,…,6 вы можете учитывать только значения y до момента времени T включительно.


<li>Выбранными моделями постройте для каждой географической зоны и каждого конца истории от 2016.04.30 23:00 до 2016.05.31 17:00 прогнозы на 6 часов вперёд; посчитайте в ноутбуке ошибку прогноза по следующему функционалу:
Qmay=1R∗739∗6∑r=1R∑T=2016.04.3023:002016.05.3117:00∑i=16y^T|T+ir−yT+ir.
Убедитесь, что ошибка полученных прогнозов, рассчитанная согласно функционалу Q, определённому на прошлой неделе, уменьшилась по сравнению с той, которую вы получили методом индивидуального применения моделей ARIMA. Если этого не произошло, попробуйте улучшить ваши модели.

<li>Итоговыми моделями постройте прогнозы для каждого конца истории от 2016.05.31 23:00 до 2016.06.30 17:00 и запишите все результаты в один файл в формате geoID, histEndDay, histEndHour, step, y. Здесь geoID — идентификатор зоны, histEndDay — день конца истории в формате id,y, где столбец id состоит из склеенных через подчёркивание идентификатора географической зоны, даты конца истории, часа конца истории и номера отсчёта, на который делается предсказание (1-6); столбец y — ваш прогноз.

<li>Загрузите полученный файл на kaggle: https://inclass.kaggle.com/c/yellowtaxi. Добавьте в ноутбук ссылку на сабмишн.

<li>Загрузите ноутбук в форму.

Подгружаем данные

In [2]:
# id нужных регионов
regsDf = pd.read_csv('../crowdRegs.csv',names=['id','regId']);  

# времянные ряды для этих регионов
df = pd.read_pickle('../loadData/crowdRegs3.pcl')
regNames = regsDf.regId.values.astype('str')
df.columns = regNames

Наверное, оптимальный способ - пройтись по всем регионам, сформировать требуемую выборку, а потом - состыковать. 
Вероятно, в процессе работы получится векторизовать это действие.
Пожалуй, имеет смысл сначала для всего фрейма добавить общие для всех колонок признаки (тренд, гармоники, даты, дни недели)

In [3]:
def processDataFrame(inpDf, Kw = 5, Ka = 2):
    """
    Обрабатываем сразу весь dateFrame и добавляем признаки, общие для всех рядов
    тренд, гармоники, категориальные перемнные
    для дат, дней недели, etc)

    Parameters:
    Kw number of weeks harmonics
    Ka number of annual harmonics
    """

    inpDf = inpDf.assign(linear = (inpDf.index - datetime.datetime(2014,1,1,0,0,0))/np.timedelta64(1, 'h'))
    
    # час — эти признаки можно пробовать брать и категориальными
    # и непрерывными, можно даже и так, и так

    # добавляем гармонические фичи
    for ind in range(1,Kw+1):
        inpDf['weekCos'+str(ind)]= np.cos(np.pi*inpDf.linear*ind/168)
        inpDf['weekSin'+str(ind)]= np.sin(np.pi*inpDf.linear*ind/168)
     
    for ind in range(1,Ka+1):
        inpDf['yearCos'+str(ind)]= np.cos(2*np.pi*inpDf.linear*ind/8766)        
        inpDf['yearSin'+str(ind)]= np.sin(2*np.pi*inpDf.linear*ind/8766)

    # добавляем числовое и категориальные свойства для дней недели
    inpDf = inpDf.assign(dayOfWeek = inpDf.index.dayofweek)
    lbDays = preprocessing.LabelBinarizer()
    lbDays.fit(list(np.arange(6)))
    DoW = pd.DataFrame(lbDays.transform(inpDf.index.dayofweek),columns = ['dayOfWeek_'+str(x) for x in np.arange(6)],
                       index = inpDf.index)      
    inpDf = inpDf.merge(DoW,left_index=True,right_index=True)

    # добавляем dummy variables для месяца
    inpDf = inpDf.assign(month = inpDf.index.month)
    lbMonths = preprocessing.LabelBinarizer()
    lbMonths.fit(list(np.arange(12)))
    Months = pd.DataFrame(lbMonths.transform(inpDf.index.month),columns = ['month_'+str(x) for x in np.arange(1,13)],
                          index = inpDf.index)      
    inpDf = inpDf.merge(Months,left_index=True,right_index=True);

    # добавляем год (вещественный)
    inpDf = inpDf.assign(year = inpDf.index.year)

    # добавляем день месяца (вещественный)
    inpDf = inpDf.assign(day = inpDf.index.day)

    # добавляем час (вещественный и категориальный)
    inpDf = inpDf.assign(hour = inpDf.index.hour)
    lbHours = preprocessing.LabelBinarizer()
    lbHours.fit(list(np.arange(24)))
    Hours = pd.DataFrame(lbHours.transform(inpDf.index.hour),columns = ['hour_'+str(x) for x in np.arange(24)],
                       index = inpDf.index)      
    inpDf = inpDf.merge(Hours,left_index=True,right_index=True)
    
    return inpDf

Теперь делаем индивидуальную обработку для каждого региона
<ol>
<li> добавляем идентификатор географической зоны — категориальный
<li> количество поездок из рассматриваемого района в моменты времени yT,yT−1,…,yT−K (параметр K можно подбирать; попробуйте начать, например, с 6)
<li> количество поездок из рассматриваемого района в моменты времени yT−24,yT−48,…,yT−24∗Kd (параметр Kd можно подбирать; попробуйте начать, например, с 2)
<li>суммарное количество поездок из рассматриваемого района за предшествующие полдня, сутки, неделю, месяц 
2) 
</ol>

In [4]:
def processSeries(df,tReg,Kh = 6, Kp = 2):
    """
    Обработка одного данного ряда 
    parameters:
        df - начальный датафрейм, из которого выберем для обработки один ряд
        tReg - название ряда, который надо обработать
        Kh - количество отслеживаемых прошлых суточных лагов "назад"
        Kp - количество отслеживаемых прошлых периодических лагов (период 24 часа)

    """
 

    tDf = df.loc[:,tReg.split() + commonFeatures].rename(columns={tReg:'y'})

    tDf = tDf.assign(region = tReg)

    for timeLag in np.arange(1,Kh+1):
        name = 'hourLag_'+str(timeLag)
        tDf.loc[:,name] = tDf.y.shift(periods=timeLag)

    for timeLag in np.arange(1,Kp+1):
        name = 'periodicLag_'+str(timeLag)
        tDf.loc[:,name] = tDf.y.shift(periods=timeLag*24)

    tDf.fillna(0,inplace=True)    

    # суммарное количество поездок из рассматриваемого района за предшествующие полдня, сутки, неделю, месяц
    tDf.loc[:,'sum12'] = tDf.y.rolling(window = 12, min_periods = 1).sum()
    tDf.loc[:,'sum24'] = tDf.y.rolling(window = 24, min_periods = 1).sum()
    tDf.loc[:,'sumWeek'] = tDf.y.rolling(window = 168, min_periods = 1).sum()
    tDf.loc[:,'sumMonth'] = tDf.y.rolling(window = 720, min_periods = 1).sum()
    
    #создаём шесть целевые переменных для каждого конца истории
    for targetVar in np.arange(1,7):
        name = 'y'+str(targetVar)
        tDf.loc[:,name] = tDf.y.shift(-targetVar)
    tDf.fillna(0,inplace=True)
    
    return tDf


In [5]:
def getTrips(X):
    model1 = bestModels.get('y1')
    model2 = bestModels.get('y2')
    model3 = bestModels.get('y3')
    model4 = bestModels.get('y4')
    model5 = bestModels.get('y5')
    model6 = bestModels.get('y6')
    pr1 = model1.predict(X)
    pr2 = model2.predict(X)
    pr3 = model3.predict(X)
    pr4 = model4.predict(X)
    pr5 = model5.predict(X)
    pr6 = model6.predict(X)
    pr1[pr1<0] = 0
    pr2[pr2<0] = 0
    pr3[pr3<0] = 0
    pr4[pr4<0] = 0
    pr5[pr5<0] = 0
    pr6[pr6<0] = 0
    
    return np.array([pr1, pr2, pr3, pr4, pr5, pr6])

In [6]:
def saveResults(rdf, fName):
    rnd = np.round

    f = open(fName,'w')
    f.writelines('id,y\n')

    for ind, row in rdf.iterrows():
        historyStart = row.date - datetime.timedelta(hours = 1)

        if historyStart > datetime.datetime(2016,6,30,17):
            continue

        s0 = str(row.region)+'_'+ str(datetime.datetime.strftime(historyStart, "%Y-%m-%d"))+ '_'+ str(historyStart.hour)

        s1 = s0 +'_1,'+str(rnd(row.get('y1'))) + '\n'
        f.writelines(s1)

        s2 = s0 +'_2,'+str(rnd(row.get('y2'))) + '\n'
        f.writelines(s2)

        s3 = s0 +'_3,'+str(rnd(row.get('y3'))) + '\n'
        f.writelines(s3)

        s4 = s0 +'_4,'+str(rnd(row.get('y4'))) + '\n'
        f.writelines(s4)

        s5 = s0 +'_5,'+str(rnd(row.get('y5'))) + '\n'
        f.writelines(s5)

        s6 = s0 +'_6,'+str(rnd(row.get('y6'))) + '\n'
        f.writelines(s6)

    f.close()    

In [67]:
# общая обработка данных
df2 = processDataFrame(df,Kw = 7, Ka = 2)
commonFeatures =  list(set(df2.columns)-set(df.columns.values))
# обработка отдельных рядов
df3 = pd.DataFrame()
for regName in regNames:
    df3 = pd.concat([df3, processSeries(df2,regName,Kh = 12, Kp = 4)])

regDf = df3.get('region')
df3 = pd.get_dummies(df3,'region')
df3 = df3.assign(region = regDf)
df3.head()

y  weekSin1  weekSin2  month  hour_19  hour_9  hour_8  \
2014-01-01 00:00:00   87  0.000000  0.000000      1        0       0       0   
2014-01-01 01:00:00   92  0.018699  0.037391      1        0       0       0   
2014-01-01 02:00:00  108  0.037391  0.074730      1        0       0       0   
2014-01-01 03:00:00   77  0.056070  0.111964      1        0       0       0   
2014-01-01 04:00:00   47  0.074730  0.149042      1        0       0       0   

                     year  hour_5  hour_4   ...    region_1684  region_1733  \
2014-01-01 00:00:00  2014       0       0   ...              0            0   
2014-01-01 01:00:00  2014       0       0   ...              0            0   
2014-01-01 02:00:00  2014       0       0   ...              0            0   
2014-01-01 03:00:00  2014       0       0   ...              0            0   
2014-01-01 04:00:00  2014       0       1   ...              0            0   

                     region_1734  region_1783  region_2068  region_2069  \
2014-01-01 00:00:00            0            0            0            0   
2014-01-01 01:00:00            0            0            0            0   
2014-01-01 02:00:00            0            0            0            0   
2014-01-01 03:00:00            0            0            0            0   
2014-01-01 04:00:00            0            0            0            0   

                     region_2118  region_2119  region_2168  region  
2014-01-01 00:00:00            0            0            0    1075  
2014-01-01 01:00:00            0            0            0    1075  
2014-01-01 02:00:00            0            0            0    1075  
2014-01-01 03:00:00            0            0            0    1075  
2014-01-01 04:00:00            0            0            0    1075  

[5 rows x 196 columns]

In [ ]:
df3.to_pickle('df3.pcl')

In [50]:
df3 = pd.read_pickle('df3.pcl')

In [68]:
df3.columns.values

array(['y', 'weekSin1', 'weekSin2', 'month', 'hour_19', 'hour_9', 'hour_8',
       'year', 'hour_5', 'hour_4', 'hour_7', 'hour_6', 'hour_1', 'hour_0',
       'hour_3', 'hour_2', 'weekCos7', 'weekCos6', 'weekCos5', 'weekCos4',
       'weekCos3', 'weekCos2', 'weekCos1', 'weekSin6', 'weekSin7',
       'weekSin4', 'weekSin5', 'month_12', 'weekSin3', 'month_10',
       'month_11', 'month_4', 'month_5', 'month_6', 'month_7', 'month_1',
       'month_2', 'month_3', 'yearSin1', 'yearSin2', 'month_8', 'month_9',
       'hour_22', 'hour_23', 'dayOfWeek', 'linear', 'day', 'hour',
       'yearCos1', 'yearCos2', 'hour_21', 'hour_20', 'hour_18',
       'dayOfWeek_5', 'dayOfWeek_4', 'dayOfWeek_3', 'dayOfWeek_2',
       'dayOfWeek_1', 'dayOfWeek_0', 'hour_11', 'hour_10', 'hour_13',
       'hour_12', 'hour_15', 'hour_14', 'hour_17', 'hour_16', 'hourLag_1',
       'hourLag_2', 'hourLag_3', 'hourLag_4', 'hourLag_5', 'hourLag_6',
       'hourLag_7', 'hourLag_8', 'hourLag_9', 'hourLag_10', 'hourLag_11',
  

Разбейте каждую из шести выборок на три части:
<ul>
<li> Обучающая, на которой будут настраиваться параметры моделей — всё до апреля 2016
<li> Тестовая, на которой вы будете подбирать значения гиперпараметров — май 2016
<li> Итоговая, которая не будет использоваться при настройке моделей вообще — июнь 2016
</ul>

Теперь надо обучить регрессор на колонке y, откинув y1-y6

Для повторяемости стоит зафиксировать random_state 

Попробовать AdaBoost Regressor над деревьями.

In [69]:
startTrain = '2014-01-01 00:00:00'
endTrain   = '2016-04-30 23:00:00'

startValidation = '2016-05-01 00:00:00'
endValidation   = '2016-05-31 23:00:00'

startTest = '2016-06-01 00:00:00'
endTest   = '2016-06-30 23:00:00'

In [70]:
targetList = ['y1','y2','y3','y4','y5','y6']
dropList = targetList
dropList.append('y')
dropList

['y1', 'y2', 'y3', 'y4', 'y5', 'y6', 'y']

In [ ]:
trainSet = df3.query('index >= @startTrain and index <= @endValidation') 
#validationSet = df3.query('intrainSetdex >= @startValidation and index <= @endValidation')
testSet = df3.query('index >= @startTest and index <= @endTest')

In [74]:
tsGroups = df3.groupby('region')

KeyError: '1371'

In [75]:
tsGroups.groups

{'1075': DatetimeIndex(['2014-01-01 00:00:00', '2014-01-01 01:00:00',
                '2014-01-01 02:00:00', '2014-01-01 03:00:00',
                '2014-01-01 04:00:00', '2014-01-01 05:00:00',
                '2014-01-01 06:00:00', '2014-01-01 07:00:00',
                '2014-01-01 08:00:00', '2014-01-01 09:00:00',
                ...
                '2016-06-30 14:00:00', '2016-06-30 15:00:00',
                '2016-06-30 16:00:00', '2016-06-30 17:00:00',
                '2016-06-30 18:00:00', '2016-06-30 19:00:00',
                '2016-06-30 20:00:00', '2016-06-30 21:00:00',
                '2016-06-30 22:00:00', '2016-06-30 23:00:00'],
               dtype='datetime64[ns]', length=21888, freq=None),
 '1076': DatetimeIndex(['2014-01-01 00:00:00', '2014-01-01 01:00:00',
                '2014-01-01 02:00:00', '2014-01-01 03:00:00',
                '2014-01-01 04:00:00', '2014-01-01 05:00:00',
                '2014-01-01 06:00:00', '2014-01-01 07:00:00',
                '2014-01-01 08

In [80]:
regId = '2119'
ts = tsGroups.get_group(regId)

In [81]:
if 'regressor1' in ts.columns.values:
    ts.drop('regressor1',axis=1,inplace=True)
    
linReg = linear_model.Ridge()
linReg.fit(ts.loc[startTrain:endTrain,:].drop(dropList,axis = 1),ts.loc[startTrain:endTrain,'y'])
print linReg.score(ts.loc[startTest:endTest,:].drop(dropList,axis = 1),ts.loc[startTest:endTest,'y'])
ts = ts.assign(regressor1 = linReg.predict(ts.drop(dropList,axis = 1)))

0.999999999999


In [ ]:
plt.figure(figsize=[15,10])
ts.y.plot()+1
ts.regressor1.plot()
plt.xlim(['06-2016','07-2016'])

In [ ]:
treeReg = DTR(random_state=42)
treeReg.fit(trainSet.drop(dropList,axis = 1),trainSet.y)
regression1 = treeReg.predict(df3.drop(dropList,axis = 1))
#df3 = df3.assign(regressor1 = regression1)
plt.plot(regression1)
print treeReg.score(testSet.drop(dropList,axis = 1),testSet.y)

In [ ]:
print linReg.score(df3.loc[startTest:endTest,:].drop(dropList,axis = 1),df3.loc[startTest:endTest,'y'])

In [ ]:
df3 = df3.assign(regressor1 = regression1)

Теперь надо обучить SARIMAX модель для каждого ряда и получить предсказания.
Для тестирования выбираем ряд 1075. Код SARIMAX модели берём из 4й недели.

In [ ]:
# получаем параметры SARIMAX модели
params = sarimaxParams.get(regId)
params = (1,1,1,1)

In [2]:
ps = range(1, 5)
d  = 1
qs = range(1, 5)

Ps = range(1, 3)
D  = 1
Qs = range(1, 3)

pList = list(product(ps, qs, Ps, Qs))

In [ ]:
endog = ts.loc[startTrain:endValidation,'y']
exog = ts.loc[startTrain:endValidation,'regressor1']

In [ ]:
results = []
best_aic = float("inf")

for params in pList:
    print params
    try:
        mSARIMA=sm.tsa.statespace.SARIMAX(ts.loc[startTrain:endValidation,'y'], order=[params[0], 2, params[1]],
                                  seasonal_order=(params[2], 1, params[3], 24),
                                  exog = ts.loc[startTrain:endValidation,'regressor1']).fit(disp=1)
    except Exception as inst:
        print inst           # __str__ allows args to be printed directly
        continue 
    else:     
        aic = mSARIMA.aic
        print('AIC',aic) 
        #сохраняем лучшую модель, aic, параметры
        if aic < best_aic:
            best_model = mSARIMA
            best_aic = aic
            best_param = param
        results.append([param, mSARIMA.aic])

In [ ]:
endog = ts.loc[:,'y']
exog = ts.loc[:,'regressor1']

plt.figure(figsize=[20,10])
endog.plot()
exog.plot()
plt.xlim(['05-2016','07-2016'])

In [ ]:
plt.plot(endog-exog)

In [ ]:
a[np.isnan(a)]

In [ ]:
a[~np.isfinite(a)]

In [ ]:
print 'Learn SARIMAX'
try:
    mSARIMA=sm.tsa.statespace.SARIMAX(ts, order=[params[0], 2, params[1]],
                                      seasonal_order=(params[2], 1, params[3], 24),
                                      exog = r_pr, enforce_invertibility = True).fit(disp=1);
except Exception as inst:
    print type(inst)     
    print inst          

# получаем предсказания регрессора на весь диапазон дат (обучение+предсказание)
exog = getRegressor(regressor,startFit,endPrediction)
# получаем данные о поездкахы на весь диапазон дат
endog = df.loc[startFit:endPrediction,tsId]

# создаём новую модель, которую будет использовать для предсказания
# Для чего такой финт ушами - не понимаю до сих пор
try:
    model_fitted = sm.tsa.statespace.SARIMAX(endog, order=[params[0], 1, params[1]],
                                         seasonal_order=(params[2], 1, params[3], 24),
                                         exog = exog).filter(mSARIMA.params)
except Exception as inst:
    print 'Can not create the model'
    print inst
    continue

# проходим по всему диапазону дат предсказаний
print 'Make prediction'
for firstLag in predictionRange[:-5]:
    lastLag = firstLag+datetime.timedelta(hours = 5)
    # prediction
    try:
        predicted_data = model_fitted.predict(firstLag, lastLag, dynamic=True, exog = exog[firstLag:lastLag])
    except Exception as inst:
        print 'Prediction error'
        print inst

Так мы получили регрессионные предсказания для всех регионов на весь доступный диапазон дат.

Выберите вашу любимую регрессионную модель и настройте её на каждом из шести наборов данных, подбирая гиперпараметры на мае 2016. Желательно, чтобы модель:
<ul>
<li>Допускала попарные взаимодействия между признаками
<li>Была устойчивой к избыточному количеству признаков (например, использовала регуляризаторы)
</ul>

Попробуем обучить Ridge классификатор с L2 регуляризацией.

In [ ]:

paramsAlpha = np.linspace(0.1,1,5)

In [ ]:
bestModels = dict()
for target in targetList:
    print target
    dropCols = [x for x in targetList if x != target ]
    dropCols.append('y')
    dropCols.append('index')
    tmpTrain = trainSet.drop(dropCols,axis = 1)
    tmpTrain.rename(columns={target:'y'},inplace=True)
    
    tmpValidation = validationSet.drop(dropCols,axis = 1)
    tmpValidation.rename(columns={target:'y'},inplace=True)
    
    minErr = np.inf
    bestAlpha = 0
    
    
    for a in paramsAlpha:
        regressor = linear_model.Ridge(alpha= a)
        regressor.fit(tmpTrain.drop('y',axis = 1),tmpTrain.loc[:,'y'])
        prediction = regressor.predict(tmpValidation.drop('y',axis = 1))
        err = np.abs(prediction - tmpValidation.loc[:,'y'])
        err = err.mean()
        print a, err
        if err <minErr:
            minErr = err
            bestAlpha = a
            bestModel = regressor
    
    bestModels.update({target:bestModel})
            
    print 'Smallest error {:2.3f} at a = {:2.3f}'.format(err,a)     

In [ ]:
# now there are three regressors. WE have to use them.
#np.save('bestModels.npy',bestModels)
#bestModels = np.load('bestModels.npy').item()

In [ ]:
bestModels = dict()


for target in targetList:
    print target
    dropCols = [x for x in targetList if x != target ]
    dropCols.append('y')
    dropCols.append('index')
    tmpTrain = trainSet.drop(dropCols,axis = 1)
    tmpTrain.rename(columns={target:'y'},inplace=True)
    
    tmpValidation = validationSet.drop(dropCols,axis = 1)
    tmpValidation.rename(columns={target:'y'},inplace=True)
    
    regressor = DTR()
    regressor.fit(tmpTrain.drop('y',axis = 1),tmpTrain.loc[:,'y'])
    prediction = regressor.predict(tmpValidation.drop('y',axis = 1))
    err = np.abs(prediction - tmpValidation.loc[:,'y'])
    err = err.mean()

    bestModels.update({target:regressor})
            
    print 'Smallest error {:2.3f} at a = {:2.3f}'.format(err,a)     

In [ ]:
dropCols = ['index','y','y1','y2','y3','y4','y5','y6']
prediction = getTrips(testSet.drop(dropCols, axis = 1))

predictionDf = pd.DataFrame(prediction.T,columns=['y1','y2','y3','y4','y5','y6'])
predictionDf.set_index(testSet.index,inplace=True)
predictionDf = predictionDf.merge(regDf,left_index=True,right_index=True,how='left')
predictionDf = predictionDf.round()

diff  = np.abs(predictionDf.y1-testSet.y1)+np.abs(predictionDf.y2-testSet.y2)+np.abs(predictionDf.y3-testSet.y3)+np.abs(predictionDf.y4-testSet.y4)+np.abs(predictionDf.y5-testSet.y5)+np.abs(predictionDf.y6-testSet.y6)
print 'Error is', diff.mean()/6

In [ ]:
# теперь надо сохранить это в файл
fName = 'res_week5-4.csv'
saveResults(predictionDf,fName)

In [ ]:
predictionDf

In [ ]:
testSet.query("region == '2168'")

deal with the dictionary of parameters

In [ ]:
np.save('sarimaxParams', sarimaxParams)  

In [ ]:
sarimaxParams